### DISCUSSION TITLE
How I can render (console logs) browser errors, warnings in spec report.

### DISCUSSION THREAD

**Discussion by bhavya-shukla-lrn** [webdriverio/webdriverio#9204] [Q&A]
I want to render errors and warnings from browser , to spec report , want the test to fail if any browser error occurs, I tried browser.getLogs but its deprecated in webdriver v7, have tried adding addConsoleLogs in wdio.conf.js reporters https://webdriver.io/docs/spec-reporter/#addconsolelogs but no luck seeing errors in specs reports, Any suggession.

**Comment by christian-bromann** [comment]
With WebdriverIO `v8` and its Bidi capabilities you will be able to create a listener for log events which will work in latest Chrome and Firefox browser, eventually Safari too. You can do this already today too, see this gist: https://gist.github.com/christian-bromann/908ae52c6b7c8f2e738a2287c2162bb6. However it is a bit cumbersome and the interface will be much easier to use in `v8`.

**Reply by th0car** [reply]
Without DevTools service and browser.getLogs() deprecated I cant find any supported/suggested way to check the browser log for errors from a WebdriverIO-test. 
1. Is the pre v8 example from 2022 above still relevant?
2. Are there any other (undocumented) ways to just get an error count from the console in any browser?

**Reply by christian-bromann** [reply]
@th0car here is an example:

```ts
await browser.sessionSubscribe({ events: ['log.entryAdded'] })

/**
 * returns: {"type":"console","method":"log","realm":null,"args":[{"type":"string","value":"Hello Bidi"}],"level":"info","text":"Hello Bidi","timestamp":1657282076037}
 */
browser.on('log.entryAdded', (entryAdded) => console.log('received %s', entryAdded))
```

Where do you think we should add this in the docs?

**Comment by th0car** [comment]
Thank you Christian!

Perhaps there should be a separate top-level menu option for a "Bidi Use Case Gallery"? This would keep the core Bidi command documentation clean and agnostic regarding how the commands can be practically applied. 

The motivation for this suggestion is that the range of practical uses for Bidi commands is difficult to predict or fully capture. It is also challenging for someone new to Bidi to quickly understand how to apply the commands in a spec, especially if they are also unfamiliar with WebdriverIO.

These examples should be small and context-independent, which makes boilerplate repositories unsuitable for this purpose.

Each example could include the following sections:

- Heading
- Use Case Area: For instance, "Client Logging," "Emulation," "Authentication," "Interactions."
- Description (The "problem" being solved. Any legacy protocol/command that could be replaced (if applicable) etc)
- The specific Bidi command used.
- Code Example: A keep it simple sample from a spec file.

This setup would likely result in the same Bidi command appearing in multiple use case areas, and some areas containing examples with different Bidi commands. While this differs from the structure of the rest of the documentation, it helps showcase how the commands can be applied in practical scenarios, making them more accessible to users.

**Reply by christian-bromann** [reply]
@th0car I [added a new page](https://github.com/webdriverio/webdriverio/pull/14040) to the best practices docs. Would love to see this section extend over time.
